### PR TITLE
Multiple images for eval.run_llava

### PR DESCRIPTION
**Pull request by HireTheHero**
PR for #57 and #197.
In case of multiple images, `eval.run_llava` with this PR could handle with arguments like below.
```
# example for two images
python -m llava.eval.run_llava \
    --model-path <model-path> \
    --image-file <image1-path><separator><image2-path> \
    --sep <separator> \
    --query "text <image-placeholder> text <image-placeholder> text"
```
My formatter added some additional changes, edit it if you'd like.

### CODE REVIEW

File: llava/eval/run_llava.py [hunk 1]
```diff
@@ -1,42 +1,81 @@
...
-        image = Image.open(BytesIO(response.content)).convert('RGB')
+        image = Image.open(BytesIO(response.content)).convert("RGB")
     else:
-        image = Image.open(image_file).convert('RGB')
+        image = Image.open(image_file).convert("RGB")
     return image
 
 
+def load_images(image_files):
+    if type(image_files) is list:
```
Comment (haotian-liu) [inline llava/eval/run_llava.py:46]: maybe redundant IFs, see above

File: llava/eval/run_llava.py [hunk 2]
```diff
@@ -1,42 +1,81 @@
...
 from PIL import Image
 
 import requests
 from PIL import Image
 from io import BytesIO
+import re
+
+
+def image_parser(args):
+    if args.sep in args.image_file:
```
Comment (haotian-liu) [inline llava/eval/run_llava.py:29]: these IFs that use two pipelines for processing single-image and multiple image conversation, seems unnecessary as the single image is just a special case of the multiple images, and "<img_path>".split("") will just return a list that has a single element, no?

**Review by haotian-liu — CHANGES_REQUESTED** [review]
Thank you for the great work, and sorry for the late response (for some reason this PR does not appear in my inbox).

Having multiple image support would be definitely beneficial. Left some comments in the code that we can discuss.

Also, do you have any examples of the model inference results so that we can just get a sense of how the model is performing. Thanks.

File: llava/eval/run_llava.py [hunk 3]
```diff
@@ -72,26 +124,33 @@ def eval_model(args):
...
 
+
 if __name__ == "__main__":
     parser = argparse.ArgumentParser()
     parser.add_argument("--model-path", type=str, default="facebook/opt-350m")
     parser.add_argument("--model-base", type=str, default=None)
     parser.add_argument("--image-file", type=str, required=True)
     parser.add_argument("--query", type=str, required=True)
     parser.add_argument("--conv-mode", type=str, default=None)
+    parser.add_argument("--sep", type=str, default="__sepsep__")
```
[the remark this answers is not in the document]
Comment (HireTheHero) [reply llava/eval/run_llava.py]: Was easy for me to debug but I guess you're right. 
Now it should look like: `parser.add_argument("--sep", type=str, default=",")`

File: llava/eval/run_llava.py (hunk 1 above)
Comment (HireTheHero) [reply llava/eval/run_llava.py]: Removed if condition to return list anyway

File: llava/eval/run_llava.py (hunk 2 above)
Comment (HireTheHero) [reply llava/eval/run_llava.py]: Now it's like this:
```
def load_images(image_files):
    out = []
    for image_file in image_files:
        image = load_image(image_file)
        out.append(image)
    return out
```

**Comment by HireTheHero** [comment]
@haotian-liu Thank you for your feedback. Requested changes should be reflected on the up2date version.

I found some good-looking example, but is too task-specific to show here.
As for general multiple-image comprehension, I admit it's quite yet-to-be-improved like this
multiple image example
(although I can see the model is trying to describe the second image, I also can see it's mixed)

For full notebook outputs, please see [here](https://github.com/HireTheHero/Machine_Notebooks/blob/main/LLaVA_multiple_image.ipynb).
For some reason you added ipynb to .gitignore, so I've temporarily added to my own repo.

I guess we have three options-
1. Pend this feature until any of us can get a good example
    - I wouldn't choose this since many users seem to be interested (maybe someone can find a good example later to show here? but it's your choice, of course)
3. Add this as nightly feature
    - Maybe reasonable. If you agree, I will rename PR's title like `[Nightly] Multiple images for eval.run_llava`
4. More investigation
    - If you have some ideas for the quick improvement (e.g. preprocessing step should be divided for multiple images), I can test it

What do you think?

**Comment by HireTheHero** [comment]
@haotian-liu Maybe LLaVA 1.5 can handle it with a one-to-one fashion prompt?
image
Full output: 
`The image features a red toy animal, possibly a toy horse, with flames coming out of its back. The toy is placed on a surface, possibly a table, and is wearing glasses. The overall appearance of the toy is unique and eye-catching, with the combination of the red color, flames, and glasses creating a distinctive and interesting visual effect.`

**Review by haotian-liu — APPROVED** [review]

**Comment by haotian-liu** [comment]
Thank you!

**Comment by shure-dev** [comment]
Hi, thank you for your work including the code update
In conclusion, do we have to wait for multi-image support until the next pre-trained model comes up?

**Comment by HireTheHero** [comment]
@shure-dev 
The example above says LLaVA 1.5 `liuhaotian/llava-v1.5-13b` might work. Appreciate it if you would test more.

**Comment by shure-dev** [comment]
@HireTheHero @haotian-liu 
Thank you 
I'm having my experiment now with a quantized model from here https://huggingface.co/4bit
but it's not working well, the model tries to recognize only the first image.

Also, I found that, in the Gradio web interface, every time I upload a second image, it clears all chat history. 
Do you know about this?

**Comment by HireTheHero** [comment]
@shure-dev Thanks for letting know of your findings. Am also personally working on LoRA/QLoRA - will keep you posted.
@haotian-liu Do you have any ideas about Gradio issue? Never used myself

**Comment by shure-dev** [comment]
Thank you for your support. I appreciate it.

I'm also wondering if it is possible for the LLaVA model to understand multiple images separately at once if we think of its architecture. Specifically, how does the model know that "first image" (text) is related to "first image" (actual image) and "second image" (text) is related to "second image" (actual image)?

**Comment by HireTheHero** [comment]
@shure-dev 
Apologies for the late response. That's an interesting question.
My example doesn't use the text "1st" or "2nd", and it gives the two images in separate user-agent interaction, I'm not sure about the model comprehension.
Maybe the model learns one-to-one text-image mapping during contrastive training, while it grasps the meaning of "first" and "second" during language training, but needs careful prompt design to test it since one-shot representation didn't work.

**Pull request merged**